modules/foreign-state-oracle/src/lib.rs

```diff
@@ -0,0 +1,229 @@
+		type Currency: NamedReservableCurrency<
+			Self::AccountId,
+			Balance = Balance,
+			ReserveIdentifier = ReserveIdentifier,
+		>;
+
+		/// The foreign state oracle module id, keeps expired queries deposits
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
+
+		/// Fee to be paid to oracles for servicing query
+		#[pallet::constant]
+		type QueryFee: Get<Balance>;
+
+		/// Timeout for query requests
+		#[pallet::constant]
+		type QueryDuration: Get<Self::BlockNumber>;
+
+		/// Dispatchable task that needs to be verified by oracle for dispatch
+		type VerifiableTask: Parameter
+			+ Dispatchable<Origin = <Self as Config>::Origin, PostInfo = PostDispatchInfo>
+			+ From<frame_system::Call<Self>>
+			+ GetDispatchInfo;
+
+		/// Origin that can dispatch calls that have been verified with foreign state
+		type OracleOrigin: EnsureOrigin<<Self as frame_system::Config>::Origin>;
+	}
+
+	#[pallet::pallet]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(_);
+
+	#[pallet::origin]
+	pub type Origin = RawOrigin;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		CreateActiveQuery { index: QueryIndex },
+		CallDispatched { task_result: DispatchResult },
+		StaleQueryRemoved { query_index: QueryIndex },
+	}
+
+	/// Index of Queries, each query gets unique number
+	#[pallet::storage]
+	#[pallet::getter(fn query_index)]
+	pub(super) type QueryCounter<T: Config> = StorageValue<_, QueryIndex, ValueQuery>;
+
+	///  The tasks to be dispatched with foriegn chain state
+	#[pallet::storage]
+	#[pallet::getter(fn active_query)]
+	pub(super) type ActiveQuery<T: Config> = StorageMap<_, Identity, QueryIndex, VerifiableCallOf<T>, OptionQuery>;
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Index key does not have an active query currently
+		NoMatchingCall,
+		/// Verifiable Call is too large
+		TooLargeVerifiableCall,
+		/// Query has expired
+		QueryExpired,
+		/// Query has not yet expired
+		QueryNotExpired,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(0)]
+		#[transactional]
+		pub fn dispatch_task(origin: OriginFor<T>, query_index: QueryIndex, data: Vec<u8>) -> DispatchResult {
+			T::OracleOrigin::ensure_origin(origin)?;
+
+			let verifiable_call = ActiveQuery::<T>::take(query_index).ok_or(Error::<T>::NoMatchingCall)?;
+			// check that query has not expired
+			ensure!(
+				verifiable_call.expiry > frame_system::Pallet::<T>::current_block_number(),
+				Error::<T>::QueryExpired
+			);
+
+			let result = verifiable_call
+				.dispatchable_call
+				.dispatch(RawOrigin::ForeignStateOracle { data }.into());
+
+			Self::deposit_event(Event::CallDispatched {
+				task_result: result.map(|_| ()).map_err(|e| e.error),
+			});
+
+			Ok(())
+		}
+
+		// TODO: Change to use idle scheduler
+		#[pallet::weight(0)]
+		#[transactional]
+		pub fn remove_expired_call(origin: OriginFor<T>, query_index: QueryIndex) -> DispatchResult {
+			ensure_none(origin)?;
```

Comment:
maybe we should make it ensure_signed or none and refund half of the query fee to the caller of this method to give some incentive for people to purge expired call?

needs to implement ValidateUnsigned to actually support unsigned tx

Reply:
yeah giving half of the fee for oracles to purge storage makes sense.